### ISSUE TITLE
[BUG] Xamarin.Forms Cannot encode SKImage as GIF

### ISSUE THREAD

**Issue by lrolleman** [mono/SkiaSharp#1130]
**Description**

When attempting to encode an SKImage as a GIF, it returns a null SKData value

Setting SKEncodedImageFormat to Png works as expected

**Code**

SKImage.FromBitmap(bitmap).Encode(SKEncodedImageFormat.Gif, 100)

**Expected Behavior**

I should get back an SKData object that encapsulates a GIF encoded image

**Actual Behavior**

I get back a null value

**Basic Information**

- Version with issue: 1.68.1.1
- Last known good version: Unknown
- IDE:  Visual Studio 2019
- Platform Target Frameworks: 
  - iOS:  13.2
  - Xamarin.Forms: 4.4.0.991537
- Target Devices:  
  - iPhone 6 (iOS 12)

**Comment by mattleibow** [comment]
Ah, yeah. Unfortunately SkiaSharp only supports 3 formats (https://github.com/mono/SkiaSharp/issues/320#issuecomment-310805723):
 - jpeg
 - png
 - webp

This is a bit unfortunate, but you could always use a separate library to do this if you really need a gif. For example (I have not actually used) there is [NGif](https://github.com/avianbc/NGif) or [BumpKit](https://github.com/DataDink/Bumpkit) or [AnimatedGif](https://www.nuget.org/packages/AnimatedGif).

A lot of them require GDI, so they might only work on Windows.

There was this discussion as well: https://groups.google.com/forum/#!searchin/skia-discuss/gif%7Csort:date/skia-discuss/R99fiu-tYNU/1G7KaxiyEAAJ
Basically, they suggest making a set of png and using ImageMagick to make a gif.

**Comment by lrolleman** [comment]
Sadly, all of those libraries rely on the Windows (System.Drawing/GDI) which just throws PlatformNotSupportedException on Xamarin. 

However, drawing inspiration from NGif (and the original Java version) and replacing all the System.Drawing components with their equivalent SkiaSharp components (Image->SKImage, Color->SKColor etc.) it just... works. Almost anti-climactic after bashing my head for a couple days on it

Performance leaves something to be desired but it's tolerable

**Comment by mattleibow** [comment]
Ooh, yeah. Do you have the code somewhere? Maybe I can take a look and see if you are maybe copying things too many times. I was actually going to have a look at this, but since you got code already, maybe we can make this an actual thing? A SkiaSharp.NGif or a NGif.SkiaSharp. People will love it. And they can now play it in Xamarin.Forms 😉

**Comment by lrolleman** [comment]
I don't have public code, though it would be very simple to do. The only part of the System.Drawing to SkiaSharp port that wasn't basic 1 to 1 was in GetImagePixels

https://github.com/avianbc/NGif/blob/master/Components/AnimatedGifEncoder.cs
Where I replaced
```
Image temp = new Bitmap(width, height);
Graphics g = Graphics.FromImage(temp);
g.DrawImage(image, 0, 0);
image = temp;
g.Dispose();
```
with 
```
var temp = new SKBitmap(width, height);
var canvas = new SKCanvas(temp);
canvas.DrawImage(image, 0, 0);
image = SKImage.FromBitmap(temp);
canvas.Dispose();
```

Fairly naive/simplistic, but works as proof of concept

**Comment by mattleibow** [comment]
Hmm. Looks good. The `new SKBitmap` is creating some memory, and then `SKImage.FromBitmap` is copying it. The `canvas.DrawImage` is also doing a copy. Is there a reason why you can't just draw the image directly?

Or, Instead of using a `SKBitmap` and then `SKImage`, just use a `SKSurface` or `SKBitmap`. In the case of GDI, the `Image` is a base type for `Bitmap`, so there is no copying going on. And, the `Graphics.FromImage` is just wrapping the image - just like the cheap `new SKCanvas`.

Hope that helps/makes sense?

**Comment by lrolleman** [comment]
That does make sense, I'll certainly do some experiments along those lines (I haven't spent much more time on it once I was convinced it was in fact possible to do)

Cheers
